lighthouse-core/audits/seo/structured-data-automatic.js

```diff
@@ -0,0 +1,173 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('../audit');
+const validateJsonLD = require('../../lib/sd-validation/sd-validation');
+const i18n = require('../../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Title of a Lighthouse audit that provides detail on whether JSON-LD structured data snippets are valid. This descriptive title is shown when no invalid JSON-LD snippets were found. */
+  title: 'JSON-LD structured data syntax is valid',
+  /** Title of a Lighthouse audit that provides detail on whether JSON-LD structured data snippets are valid. This descriptive title is shown when JSON-LD snippets with invalid content were found. */
+  failureTitle: 'JSON-LD structured data syntax is invalid',
+  /** Description of a Lighthouse audit that tells the user whether JSON-LD snippets on the page are invalid. This is displayed after a user expands the section to see more. No character length limits. */
+  /* eslint-disable-next-line max-len */
+  description: 'Structured data contains rich metadata about a web page. The data is used in search results and social sharing. Invalid metadata will affect how the page appears in these contexts. This audit is currently doing basic JSON-LD validation. See also the manual audit below to validate other types of structured data.',
+  /** Explanatory message stating what percentage of JSON-LD structured data snippets are invalid */
+  displayValue: '{validSnippetProportion, number, percent} valid snippets',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class StructuredDataAutomatic extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'structured-data-automatic',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['ScriptElements'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts) {
+    const jsonLDElements = artifacts.ScriptElements.filter(
+      scr => scr.type === 'application/ld+json' && !!scr.content);
+
+    if (jsonLDElements.length === 0) {
+      return {
+        notApplicable: true,
+        rawValue: true,
+      };
+    }
+
+    const validatedSnippets = await Promise.all(jsonLDElements.map(async (element) => {
+      // We don't want to show empty lines around the snippet
+      const content = (element.content || '').trim();
```

Comment:
bump :)

Reply:
It's `notApplicable` now because we do this:

``` javascript
const jsonLDElements = artifacts.ScriptElements.filter(
      script => script.type === 'application/ld+json' && !!script.content);
```

Have changed the `|| ''` to a `/** @type ` cast.

Reply:
we don't seem to be trimming that filter about though was what I was trying to say